designs/0019-variadic_odes.md

```diff
@@ -0,0 +1,171 @@
+4. ```times``` - Sorted arary of times to which the ode will be solved (each
+  element must be greater than t0)
+5. ```rel_tol``` - Relative tolerance for solve (data)
+6. ```abs_tol``` - Absolute tolerance for solve (data)
+7. ```max_num_steps``` - Maximum number of timesteps to take in integrating
+  the ODE solution between output time points (data)
+5. ```arg1, arg2, ...``` - Arguments passed unmodified to the ODE right hand
+  side (`f`). Can be any non-function type.
+
+There are a few changes from the previous implementation.
+
+1. `theta`, `x_real`, and `x_int` are no longer broken out. All parameters,
+real data, and integer data should be passed through as a trailing argument
+(with no rules on naming).
+
+2. The state is written in terms of a `vector` instead of real array. The
+reasoning for switching is that an ODE state is a single thing (a single
+vector), not really a list of separate things (where an array would be
+preferable). Also ODE states are usually represented as vectors mathematically.
+
+3. The times array only needs to be sorted, not strictly increasing. This means
+that there can multiple output times in a row with the same value.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation of these features is relatively straightforward given the
+code from the
+[reduce_sum](https://github.com/stan-dev/design-docs/blob/master/designs/0017-reduce_sum.md)
+design doc and the existing ODE infrastructure. There is not any significant new
+technology needed.
+
+The old `integrate_ode_bdf`, `integrate_ode_adams`, and `integrate_ode_rk45`
+functions should be deprecated with the introduction of these new functions.
+They can remain in the code to provide compatibility to models written with the
+old interfaces. These functions can all be written to wrap around the new
+interfaces so there will not be much technical debt accrued maintaining them.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+One drawback is that the existing ODE documentation and case studies will need
+to be updated to point at the new interfaces. The old interfaces can
+remain for backwards compatibility though.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The most obvious alternatives are:
+
+* Change the scope of the functions block to include variables defined in
+the `data`, `transformed data`, `parameters`, and `transformed parameters`
+block.
+
+* Wait until closures are available so that the ODE right hand side can be
+defined by a closure (which would then have the scope to use any variable
+that could otherwise be passed to through as a right hand side argument here).
+
+Neither of these are incompatible with the current design though, and so do not
+make the current proposal less attractive.
+
+# Prior art
+[prior-art]: #prior-art
+
+This builds directly on the existing ODE solver and `reduce_sum` interfaces.
```

Comment:
Mention that the old tests stay in place.

Reply:
I put this in the reference level explanation.